File: leetcode78/solution.py
class Solution:
    '''
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        subsets=[]

        # create subsets with the number at index i
        def create_subset(i):
            # if the index is equal to the length of the array
            # that means we have reached the end
            # return the array
            if i==len(nums):
                res.append(subsets[:])
                return
            subsets.append(nums[i])
            create_subset(i+1)

            # remove the last element from the subset to backtrack and
            # recursively call create_subset with the same index
            subsets.pop()
            create_subset(i+1)
        
        create_subset(0)
        return res
    '''
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        def dfs(sub, start):
            res.append(sub[:])

            for i in range(start, len(nums)):
                sub.append(nums[i])
                dfs(sub, i+1)
                sub.pop()
        dfs([],0)
        return res

File: leetcode78/Solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> sub;
        dfs(nums, 0, sub, res);
        return res;
    }

private:
    void dfs(vector<int>& nums, int start, vector<int>& sub, vector<vector<int>>& res) {
        res.push_back(sub);

        for (int i = start; i < nums.size(); i++) {
            sub.push_back(nums[i]);
            dfs(nums, i + 1, sub, res);
            sub.pop_back(); // backtrack
        }
    }
};
